tasks/install.yml

```diff
@@ -85,26 +85,39 @@
         update_cache: true
       check_mode: false
 
-- name: Ensure gitlab-rails file exist
+- name: Check if VERSION file exists
   ansible.builtin.stat:
-    path: "/usr/bin/gitlab-rails"
-  register: file
+    path: "/var/opt/gitlab/gitlab-rails/VERSION"
+  register: gitlab_version_file
 
-- name: Wait until all previous background migrations are processed.
-  become: yes
-  ansible.builtin.command:
-    cmd: gitlab-rails runner -e production 'puts Gitlab::BackgroundMigration.remaining'
-  register: remaining
-  until: remaining.stdout == "0"
-  retries: 10
-  delay: 10
-  check_mode: no
-  changed_when: remaining.rc != 0
+- name: Wait for background migrations on primary node
   when:
-    - file.stat.exists
-    - file.stat.executable
-    - gitlab_is_primary is defined
     - gitlab_is_primary
+    - gitlab_version_file.stat.exists
+    - gitlab_rails_binary.stat.exists
+    - gitlab_rails_binary.stat.executable
+  block:
+    - name: Get the currently installed GitLab version
+      ansible.builtin.slurp:
+        path: "/var/opt/gitlab/gitlab-rails/VERSION"
+      register: gitlab_version_base64
+
+    - name: Set fact about installed GitLab version
+      ansible.builtin.set_fact:
+        gitlab_installed_version: "{{ gitlab_version_base64['content'] | b64decode }}"
+      check_mode: no
+
+    - name: Wait until all previous background migrations are processed.
+      become: yes
+      ansible.builtin.command:
+        cmd: gitlab-rails runner -e production 'puts Gitlab::BackgroundMigration.remaining'
+      register: remaining
+      until: remaining.stdout == "0"
+      retries: 10
+      delay: 10
+      check_mode: no
+      changed_when: false
+      when: gitlab_version == '' or gitlab_installed_version is version(gitlab_version, operator='lt', strict=True)
```

Comment:
I understand this as follows:

* If no `gitlab_version` is defined it will install latest GitLab. In this case we tell it to always wait for background migrations to be finished
* Only if the `gitlab_installed_version` is lower than `gitlab_version`, which is to be installed, the check is performed.

If my description is correct, I think this is fine :+1:

Reply:
Exactly :ok_hand: